### DISCUSSION TITLE
Battery Stats

### DISCUSSION THREAD

**Discussion by Sboshoff76** [celsworth/lxp-bridge#195] [General]
I'd love to get a better understanding of my batteries.  mqtt explorer shows these entries for me. I looked at the register documentation as well as the EG4 modbus protocol doc but couldn't really puzzle much together. Does anyone have a clue?

Is it just my setup that has 0 values for some of these (most notably SOH and cycle count which should definitely not be 0)?

  "soh": 0, - State oh health (My 4 dyness batteries always shows 0 for this)
  "bat_status_0": 0, -  ??
  "bat_status_1": 0, -  ??
  "bat_status_2": -16384, -  ??
  "bat_status_3": 0, -  ??
  "bat_status_4": 0, -  ??
  "bat_status_5": 4608, -  ??
  "bat_status_6": -21569, -  ??
  "bat_status_7": 4, -  ??
  "bat_status_8": 6172, -  ??
  "bat_status_9": 3173, -  ??
  "bat_status_inv": 3, -  BMS status collected by inverter (what does 3 mean ?)
  "bat_count": 4,
  "bat_capacity": 400,
  "bms_event_1": 0,  -  ??
  "bms_event_2": 0, -  ??
  "max_cell_voltage": 0, -  Maximum cell voltage (My 4 dyness batteries always shows 0 for this)
  "min_cell_voltage": 0, -  Minimum cell voltage (My 4 dyness batteries always shows 0 for this)
  "max_cell_temp": 0, -  Maximum cell temp (My 4 dyness batteries always shows 0 for this)
  "min_cell_temp": 0, -  Minimum cell temp (My 4 dyness batteries always shows 0 for this)
  "bms_fw_update_state": 528,  -  EG4 doc shows 1-Upgrading 2-Upgrading successful 3-Upgrading failed
  "cycle_count": 0, - Number of charge and discharge cycles (My 4 dyness batteries always shows 0 for this)

**Comment by celsworth** [comment]
What all the individual pieces mean I'm not quite sure yet, this is something I would like to discover and document. I did make a start on this in https://github.com/celsworth/lxp-bridge/issues/7 but they're just the obvious ones really.

I can give some pointers that might help you figure some of it out.

So the first thing to bear in mind is that these are 16bit integers (they're presented here as signed hence -21569, but I think thats incorrect, they should be unsigned) that are originally sent from the inverter as two 8bit integers. And in some data from the inverter, these original 8bit integers are distinct (if related) and make sense on their own.

So for example, taking 6172 from bat_status_8, breaking this down into its 8bit constituents gives us 24 and 28. I tend to use Python's `divmod` as a quick way to do this:

```
m1p:~> python3
Python 3.11.5 (main, Aug 24 2023, 15:09:45) [Clang 14.0.3 (clang-1403.0.22.14.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> divmod(6172, 256)
(24, 28)
```

So while I have no idea what these numbers mean, its possible that 24 and 28 are more meaningful than 6172, for example. Like, this could mean 24c and 28c or something being the temperatures, who knows.

The other possibility is the individual bits in each integer all mean their own thing. So this is already known for bat_status_5 for example, when the 4th bit is set, the battery is making a full charge request (for BMS balancing etc).

I tend to use `bc` for breaking these out, other tools are available:

```
m1p:~> bc
>>> obase=2
>>> 4608
1001000000000
```

So in your case you have bit 10 and 13 set, but no others. No idea what they mean though. But if you can provoke certain behaviours in your batteries, or watch them as they cool down as we go into winter, you might see various ints/bits change and be able to associate what they are. 

`bat_status_inv` seems to be the most important data from all the other bat_status fields, although most of it is unknown, I can tell you that 3 just means charging and discharging mosfets are both enabled (first two bits are set, being 1 + 2 = 3). This is mentioned in #7. I know these because they're fairly easy to trigger to turn on and off when the battery gets full or empty.

My `soh` and `cycle_count` were always 0 too, it probably depends on having a new enough firmware and getting the correct data from the batteries over canbus.

**Reply by celsworth** [reply]
Latest dev build should fix the negative numbers, as a small improvement.
